cpmpy/transformations/linearize.py

```diff
@@ -5,190 +5,187 @@
-            return [cpm_expr._bv <= 0]
-        return [cpm_expr >= 1]
+        return [sum([cpm_expr]) >= 1]
 
-    if cpm_expr.name == "and":
-        if all(arg.is_bool() for arg in cpm_expr.args):
+    # Boolean operators
+    if isinstance(cpm_expr, Operator) and cpm_expr.is_bool():
+        # conjunction
+        if cpm_expr.name == "and":
             return [sum(cpm_expr.args) >= len(cpm_expr.args)]
-        raise Exception("Numeric constants or numeric variables not allowed as base constraint")
 
-    if cpm_expr.name == "or":
-        if all(arg.is_bool() for arg in cpm_expr.args):
+        # disjunction
+        elif cpm_expr.name == "or":
             return [sum(cpm_expr.args) >= 1]
-        raise Exception("Numeric constants or numeric variables not allowed as base constraint")
 
-    if cpm_expr.name == "->":
-        cond, sub_expr = cpm_expr.args
-        if not cond.is_bool() or not sub_expr.is_bool():
-            raise Exception(
-                f"Numeric constants or numeric variables not allowed as base constraint, cannot linearize {cpm_expr}")
-        lin_comps = linearize_constraint(sub_expr)
-        lin_exprs = []
-        for l_expr in lin_comps:
-            lhs, rhs = l_expr.args
-            if lhs.name == "mul" and _wsum_should(lhs):
-                # multiplication with var and constant, convert to wsum
-                lhs = Operator("wsum", _wsum_make(lhs))
+        # xor
+        elif cpm_expr.name == "xor" and len(cpm_expr.args) == 2:
+            return [sum(cpm_expr.args) >= 1]
 
-            if isinstance(rhs, _NumVarImpl):
-                # Vars on rhs of linexpr not supported in indicator constraints
-                if isinstance(lhs, _NumVarImpl):
-                    new_lhs = lhs + (-1) * rhs
-                elif lhs.name == "sum":
-                    new_lhs = Operator("wsum", [[1]*len(lhs.args) + [-1], lhs.args + [rhs]])
-                elif lhs.name == "wsum":
-                    new_lhs = lhs + (-1 * rhs)
-                else:
-                    raise Exception(f"Unsupported expression {lhs} on right hand side of implication {l_expr}, resulting from linearization of {cpm_expr}")
-                lin_exprs += [Comparison(l_expr.name, new_lhs, 0)]
-            else:
-                lin_exprs += [l_expr]
+        # reification
+        elif cpm_expr.name == "->":
+            # determine direction of implication
+            cond, sub_expr = cpm_expr.args
+            assert isinstance(cond, _BoolVarImpl), f"Linearization of {cpm_expr} is not supported, lhs of implication must be boolvar. Apply `only_bv_implies` before calling `linearize_constraint`"
 
-        return [cond.implies(l_expr) for l_expr in lin_exprs]
+            if isinstance(cond, _BoolVarImpl) and isinstance(sub_expr, _BoolVarImpl):
+                # shortcut for BV -> BV, convert to disjunction and apply linearize on it
+                return linearize_constraint(~cond | sub_expr, reified=reified)
```

Comment:
What if `cond` is a negation, e.g., `~p`? Do we introduce double negation then? Can `only_positive_bv` fix this?

In general, I'd expect to not introduce more negations if not needed. Perhaps this could become `linearize_constraint(sub_expr >= cond)`?

Reply:
If cond is a negated boolean variable, we'll negate it and it will become a positive variable (in the constructor).
If cond is a positive boolvar, it will become negative and be removed by only_positive_bv

Reply:
First is good. However, it would be good to have the invariant that linearize does not introduce negations: we risk getting loops, because `only_positive_bv` calls linearize again on auxiliary constraints. I know the loop does not occur in practice, but it will definitely not occur of we have this invariant.

Also, we should avoid making many auxiliary variables (which is done by `only_positive_bv`).

So let's go with the above suggestion?

Reply:
Aha I see, yes will adapt it to your suggestion.